### PR TITLE
New Shell

### PR DESCRIPTION
**Pull request by iamrecursion**
This shell allows the interpretation of both projects and stand-alone luna files. It will be integrated with the type-checker once it can report errors correctly.

### CODE REVIEW

File: core/src/OCI/Data/Name/Qualified.hs [hunk 1]
```diff
@@ -42,7 +46,7 @@ instance Convertible Qualified Name.Name where
     {-# INLINE convert #-}
 
 instance Convertible [Name.Name] Qualified where
```
Comment (wdanilo) [inline core/src/OCI/Data/Name/Qualified.hs]: it's ok, but could be more flexible. ou can define it as `Convertible [a] Qualified` with `Convertible' a Name` constraint.

File: passes/src/Luna/Pass/Sourcing/ClassProcessor.hs [hunk 2]
```diff
@@ -80,7 +80,7 @@ generateGetter :: Pass.Interface ClassProcessor m
 generateGetter consName fieldsCount fieldPos fieldName = do
     retVar   <- IR.var  fieldName
     matchVar <- IR.var' fieldName
-    selfVar  <- IR.var  $ convert "self"
+    selfVar  <- IR.var  $ convert ("self" :: String)
```
Comment (wdanilo) [inline passes/src/Luna/Pass/Sourcing/ClassProcessor.hs]: Why are you making something polymorphic less polymorphic and then converting it (which basically makes it polymorphic again)?

File: project/src/Luna/Project.hs [hunk 3]
```diff
@@ -3,20 +3,22 @@ module Luna.Project where
...
+import qualified System.FilePath      as FilePath
+import qualified System.FilePath.Find as Find
+
+import Control.Arrow           ((&&&))
+import Control.Exception.Safe  (try, tryAny)
+import Data.Bimap              (Bimap)
+import Data.Map                (Map)
+import OCI.Data.Name           (Name)
+import OCI.Data.Name.Qualified (Qualified)
+import Path                    (Path, Abs, Rel, File, Dir, (</>))
```
Comment (wdanilo) [inline project/src/Luna/Project.hs]: I strongly tend to use all of these (apart from the operator) as qualified names. The same regarding `System.getEnv` below. It just makes reading code much more easy and following things much clear.

File: project/src/Luna/Project.hs [hunk 4]
```diff
@@ -59,7 +67,7 @@ projectRootForFile :: (MonadIO m, MonadThrow m)
                    => Path Abs File -> m (Path Abs Dir)
 projectRootForFile file = do
     maybeRoot <- findProjectRoot $ Path.parent file
-    maybe (throwM (ProjectNotFoundException file)) return maybeRoot
+    maybe (throwM (ProjectNotFoundException file)) pure maybeRoot
```
Comment (wdanilo) [inline project/src/Luna/Project.hs]: Please do not use `MonadThrow`, use explicit error-tracking monad instead. It's one of the most important thing in our design.

File: project/src/Luna/Project.hs [hunk 5]
```diff
@@ -81,43 +89,46 @@ getLunaProjectsFromDir :: (MonadIO m, MonadThrow m)
...
 findProjectFile :: (MonadIO m, MonadThrow m)
                 => Path Abs Dir -> m (Maybe (Path Abs File))
 findProjectFile dir = getLunaProjectsFromDir dir >>= \case
     [] -> let parentDir = Path.parent dir in
-        if parentDir == dir then return Nothing else findProjectFile parentDir
-    [projectFile] -> return $ Just projectFile
-    _             -> return Nothing
+        if parentDir == dir then pure Nothing else findProjectFile parentDir
+    [projectFile] -> pure $ Just projectFile
+    _             -> pure Nothing
```
Comment (wdanilo) [inline project/src/Luna/Project.hs]: all these pures can b refactored out

File: shell/src/Luna/Shell/Command.hs [hunk 6]
```diff
@@ -30,99 +30,13 @@ type ConfigStateIO m =
...
-    { __packages :: [String]
-    } deriving (Eq, Generic, Ord, Show)
-makeLenses ''InstallOpts
-
-newtype DownloadOpts = DownloadOpts
-    { __packages :: [String]
-    } deriving (Eq, Generic, Ord, Show)
-makeLenses ''DownloadOpts
-
-data Command
```
Comment (wdanilo) [inline shell/src/Luna/Shell/Command.hs]: We should not remove it. We should leave visible for the end user the options that are working now, but all the options which are implemented as described in the RFC should be left here in some way, so we will not re-implement it in a moment. This is good code, should not be removed.

File: shell/src/Luna/Shell/Command.hs [hunk 7]
```diff
@@ -201,69 +82,7 @@ download opts = liftIO $ print opts
...
-            Retract  opts -> retract  opts
-            Options  opts -> options  opts
-            Rollback opts -> rollback opts
-            Update   opts -> update   opts
-            Freeze   opts -> freeze   opts
-            Unfreeze opts -> unfreeze opts
-            Install  opts -> install  opts
-            Download opts -> download opts
-
-acquireGlobalConfig :: forall m . MonadIO m => m (Either Text Global.Config)
```
Comment (wdanilo) [inline shell/src/Luna/Shell/Command.hs]: This code (and all related) should not be removed. If we have functionality of aquiring some global user config, we should leave it and develop further when needed. This global config contains some information, like users mail etc. If we already implemented it, we should not remove it.

File: package/configuration/src/Luna/Package/Configuration/Global.hs [hunk 8]
```diff
@@ -1,137 +0,0 @@
-module Luna.Package.Configuration.Global where
```
Comment (wdanilo) [inline package/configuration/src/Luna/Package/Configuration/Global.hs]: Imo, we should not remove this code. If it implements some parts of the config RFC, we should leave it and extend on demand. We put time to create it, it should be working. If its not, we should spend 1-2 days to make it working with current shell (looking at the code - thats the time that could be needed imo)

File: package/configuration/src/Luna/Package/Configuration/License.hs [hunk 9]
```diff
@@ -1,130 +0,0 @@
-module Luna.Package.Configuration.License where
```
Comment (wdanilo) [inline package/configuration/src/Luna/Package/Configuration/License.hs]: This whole module contains good code, will be needed sooner than later, should nto be deleted

File: package/configuration/test/spec/Luna/Test/Package/Configuration/LicenseSpec.hs [hunk 10]
```diff
@@ -1,53 +0,0 @@
-module Luna.Test.Package.Configuration.LicenseSpec where
```
Comment (wdanilo) [inline package/configuration/test/spec/Luna/Test/Package/Configuration/LicenseSpec.hs]: Parsing license files accordign to how RFC was designed is not only important but also not very obvious. Removing good work here is bad movement imo.

File: core/src/OCI/Data/Name/Qualified.hs (hunk 1 above)
Comment (iamrecursion) [reply core/src/OCI/Data/Name/Qualified.hs]: Nice catch. Done.

File: passes/src/Luna/Pass/Sourcing/ClassProcessor.hs (hunk 2 above)
Comment (iamrecursion) [reply passes/src/Luna/Pass/Sourcing/ClassProcessor.hs]: These were meant to be temporary to help something compile! Fixed now.

File: project/src/Luna/Project.hs (hunk 3 above)
Comment (iamrecursion) [reply project/src/Luna/Project.hs]: Fixed.

File: core/src/OCI/Data/Name/Qualified.hs (hunk 1 above)
Comment (iamrecursion) [reply core/src/OCI/Data/Name/Qualified.hs]: After having been informed by Marcin that conversions to `Name` pollute the name table, I'm sticking with the existing instance.

File: project/src/Luna/Project.hs (hunk 4 above)
Comment (iamrecursion) [reply project/src/Luna/Project.hs]: `MonadThrow` is required by `Path`. I have added a `MonadException` constraint for our own exception.

File: project/src/Luna/Project.hs (hunk 5 above)
Comment (iamrecursion) [reply project/src/Luna/Project.hs]: Unless I'm missing something, refactoring them out makes the code harder to understand.

**Pull request closed without merging**